actionpack/lib/action_dispatch/middleware/host_authorization.rb

```diff
@@ -106,17 +113,25 @@ def call(env)
       VALID_ORIGIN_HOST = /\A(#{HOSTNAME})(?::\d+)?\z/
       VALID_FORWARDED_HOST = /(?:\A|,[ ]?)(#{HOSTNAME})(?::\d+)?\z/
 
-      def authorized?(request)
-        origin_host = request.get_header("HTTP_HOST")&.slice(VALID_ORIGIN_HOST, 1) || ""
-        forwarded_host = request.x_forwarded_host&.slice(VALID_FORWARDED_HOST, 1) || ""
+      def blocked_host_if_any(request)
+        valid_origin_host = request.get_header("HTTP_HOST")&.slice(VALID_ORIGIN_HOST, 1) || ""
+        valid_forwarded_host = request.x_forwarded_host&.slice(VALID_FORWARDED_HOST, 1) || ""
 
-        @permissions.allows?(origin_host) && (forwarded_host.blank? || @permissions.allows?(forwarded_host))
+        if valid_origin_host.blank? || @permissions.block?(valid_origin_host)
+          request.get_header("HTTP_HOST")
+        elsif valid_forwarded_host.present? && @permissions.block?(valid_forwarded_host)
+          request.x_forwarded_host
+        end
       end
 
       def excluded?(request)
         @exclude && @exclude.call(request)
       end
 
+      def mark_as_blocked(request, blocked_host)
+        request.set_header("action_dispatch.blocked_host", blocked_host)
```

Comment:
I'm in doubt about setting this "action_dispatch.blocked_host" header:

- is it okay to introduce a new custom header?
- is it the right name for it?

I'd prefer to pass blocked_host to `@response_app` but it's not backward compatible since people may provide custom response apps it seems.

Another alternative would be to show both `Host` and `X-Forwarded-Host` in the error message body but I'm not sure about it either.